R/SEIRDV.R

```diff
@@ -0,0 +1,355 @@
+#' \deqn{\frac{dR(t)}{dt} = gamma I(t) - delta_R R(t)}
+#' \deqn{\frac{dV(t)}{dt} = nu Inter(t) S(t) - delta_V V(t)}
+#' \deqn{\frac{dC(t)}{dt} = beta S(t) I(t)}
+#' \deqn{\frac{dD(t)}{dt} = mu I(t)}
+#'
+#' This function relies on the package deSolve.
+#'
+#' @param object an object of the class SEIRDV
+#' @param times (double) a sequence of time points at which the solution to
+#' the system of ODEs should be returned. Must be of the form
+#' seq(t_start, t_end, by=t_step). Default time series is seq(0, 100, by = 1).
+#' @param solve_method (string) a string of chosen numerical integration method
+#' for solving the ode system. Default is lsoda which is also the default for
+#' the ode function in the deSolve package used in this function.
+#'
+#' @return two dataframes: one with the time steps, age range, time series of S,
+#' E, I, R and V population fractions, and one with the time steps, age range,
+#' time series of incidences and deaths population fraction.
+#' 
+#' @export
+#' 
+setMethod(
+  "run", "SEIRDV",
+  function(object, times, solve_method = "lsoda") {
+    if (!is.double(times)) {
+      stop("Evaluation times of the model storage format must be a vector.")
+    }
+    if (is.null(unlist(object@transmission_parameters)))
+      stop("Transmission parameters must be set before running.")
+    if (is.null(unlist(object@initial_conditions)))
+      stop("Initial conditions must be set before running.")
+    if (is.null(unlist(object@intervention_parameters)))
+      stop("Intervention parameters must be set before running.")
+    
+    # set initial state vector
+    state <- c(S = initial_conditions(object)$S0,
+               E = initial_conditions(object)$E0,
+               I = initial_conditions(object)$I0,
+               R = initial_conditions(object)$R0,
+               V = initial_conditions(object)$V0,
+               C = 0,
+               D = 0)
+    # set transmission parameters vector
+    parameters <- c(b = transmission_parameters(object)$beta,
+                    k = transmission_parameters(object)$kappa,
+                    g = transmission_parameters(object)$gamma,
+                    m = transmission_parameters(object)$mu,
+                    n = transmission_parameters(object)$nu,
+                    d_v = transmission_parameters(object)$delta_V,
+                    d_r = transmission_parameters(object)$delta_R)
+    
+    # set intervention parameters vector
+    int_parms <- 
+      InterventionParameters(
+        start=intervention_parameters(object)$starts,
+        stop=intervention_parameters(object)$stops,
+        coverage= intervention_parameters(object)$coverages)
+    
+    sim_parms <- SimulationParameters(start =0, stop = tail(times, n=1),
+                                      tstep = 0.1)
+    
+    inter_prot <- intervention_protocol(int_parms, sim_parms, 1) %>%
+      filter(time %in% times) %>%
+      .$coverage
+    
+    intervention <- approxfun(times, inter_prot, rule=2)
```

Comment:
Is there any particular reason why you are discretizing the smoothed intervention, then interpolating it, when the smoothed intervention (sums of your tanh function) can be evaluated directly at any time point? If it's due to some performance consideration, you could mention that in a comment somewhere.